### DISCUSSION TITLE
Some questions about crossover method

### DISCUSSION THREAD

**Discussion by WalterLu3** [ERGO-Code/HiGHS#1709] [Q&A]
I am currently looking at the crossover routine `Highs_crossover`, and I have some questions and wonders if anyone is aware of the answer. First of all, I would like to sum up what I observe what `Highs_crossover` is doing with an assumption that only the primal variable values are fed as input of the routine. Please correct me if I am wrong or inaccurate at any step. The following are the steps it takes:

1. Project $x$ variables (original LP problem's primal variables) to variable bounds and get $\bar{x}$. 
2. Calculate the slacks $s$ of constraints using x. If $s$  is not within its bound, move it to its closest
bound, and get $\bar{s}$. After this step, we regard $(\bar{x},\bar{s})$ as the primal variables of the LP problem in standard form.
3. Find a dual point, $(y,z)$, for the primal point we found previously. $(\bar{x},\bar{s})$
4. Check that $\bar{x} \perp y$ and $\bar{s} \perp z$
5. Start constructing a (well-conditioned) basis with some priority weights, which is related to the primal and dual values of the variables.
6. Do dual and primal push.

I have two main questions:
1. How do you construct the dual point in step 2 giving a primal solution? Should I look at it from the aspect of having a point in the primal LP problem and we want to find a corresponding point in the dual LP problem? I was stuck here because I think $(\bar{x},\bar{s})$ can be primal infeasible (i.e. they are within their variable bounds but some standard form constraints may not be satisfied).
2.  For the step 6. What are the goal of primal push and dual push? In the documentation, it says that the routine will push some of dual and primal variables to their bounds. Are they used to get rid of primal and dual infeasibility? I saw that [Gurobi also does primal and dual push](https://www.gurobi.com/documentation/current/refman/crossover.html), but I did not find any explanations of why it does that.

Thanks.

**Comment by wang123zhao551451** [comment]
Maybe you need to read more paper, E.g. Andersen, E. D. (1999). On exploiting problem structure in a basis identification procedure for linear programming. INFORMS Journal on Computing, 11(1), 95–103. https://doi.org/10.1287/ijoc.11.1.95

**Comment by jajhall** [comment]
Sorry, I missed this question six months ago. The crossover process is not something I'll explain here. Look at the paper and the code

**Reply by WalterLu3** [reply]
Thanks.
